storage/innobase/xtrabackup/src/ds_decrypt.c

```diff
@@ -355,37 +355,51 @@ decrypt_write(ds_file_t *file, const void *buf, size_t len)
 	nthreads = crypt_ctxt->nthreads;
 
 	if (crypt_file->buf_len > 0) {
+		const size_t remainder = crypt_file->buf_len;
+		size_t new_data_size = 0;
 		thd = threads;
 
 		pthread_mutex_lock(&thd->ctrl_mutex);
-
 		do {
-			if (parse_result == XB_CRYPT_READ_INCOMPLETE) {
-				crypt_file->buf_size = crypt_file->buf_size * 2;
+			if (parse_result == XB_CRYPT_READ_INCOMPLETE
+			    || crypt_file->buf_size == crypt_file->buf_len) {
+				crypt_file->buf_size *= 2;
 				crypt_file->buf = (uchar *) my_realloc(
 						crypt_file->buf,
 						crypt_file->buf_size,
 						MYF(MY_FAE|MY_ALLOW_ZERO_PTR));
 			}
 
-			memcpy(crypt_file->buf + crypt_file->buf_len,
-			       buf, MY_MIN(crypt_file->buf_size -
-					   crypt_file->buf_len, len));
+			new_data_size = MY_MIN(crypt_file->buf_size -
+					       crypt_file->buf_len, len);
```

Comment:
s/crypt_file->buf_len/remainder/ here and below?

Reply:
fixed that thing also